Flashlight-VNC/src/com/flashlight/vnc/VNCClient.as

```diff
@@ -599,78 +609,64 @@ package com.flashlight.vnc
+				rfbWriter.writeKeyEvent(false, 0xFFE1, false);
+			}
+			if (charToSend.controled == true) {
+				rfbWriter.writeKeyEvent(false, 0xFFE3, false);
+			}
+			rfbWriter.writeKeyEvent(false, charToSend.code, true);
+		}
 
-				//logger.info("<< onLocalKeyboardEvent()");
+		private function sendCharsFromQueue():void {
+			if (charsToSend.length > 0) {
+				var nextChar:Object = charsToSend[0];
+				sendChar(nextChar);
+				charsToSend.splice(0, 1);
+				setTimeout(sendCharsFromQueue, pastePauseDelay);
+			} else {
+				captureKeyEvents = true;
 			}
 		}
 
 		private function onTextInput(event:TextEvent):void {
 			if (status != VNCConst.STATUS_CONNECTED) return;
-
 			if (captureKeyEvents) {
 				logger.info(">> onTextInput()");
-				logger.info("Shift Key Down? " + shiftKeyDown);
-
-				expectKeyInput = false;
 
 				var input:String = event.text;
 
-				logger.info("event.text " + input);
-
-				var useShift:Boolean;
-				var cc:uint;
-                // :?<>"{}+_)(*&^%$#@!~
-                var needsShift:Array = [];
-				var chars:String = ":?<>|\"{}+_)(*&^%$#@!~ABCDEFGHIJKLMNOPQRSTUVWYXZ";
-				var i:int = 0;
-
-				for (i = 0; i < chars.length; i++) {
-					needsShift.push(chars.charCodeAt(i));
-				}
-
-				var waitTime:int = 0;
 				if(input.length > 1) {
+					//We are pasting some text now
 					captureKeyEvents = false;
 				}
-				for (i=0; i<input.length ;i++) {
-					cc=input.charCodeAt(i);
-					rfbWriter.writeKeyEvent(false,0xFFE3,true);
-					useShift = needsShift.indexOf(cc) >= 0;
-					if(useShift){
-						 logger.info("Using shift for char " + cc);
-						 setTimeout(rfbWriter.writeKeyEvent, waitTime += 45, false, 0xFFE3, true);
-						 setTimeout(rfbWriter.writeKeyEvent, waitTime += 5, true, 0xFFE1, true);
-					} else {
-		        setTimeout(rfbWriter.writeKeyEvent, waitTime += 15, false, 0xFFE3, true);
-		        setTimeout(rfbWriter.writeKeyEvent, waitTime += 5, false, 0xFFE1, true);
-		      }
-					setTimeout(rfbWriter.writeKeyEvent, waitTime += 15, false, 0xFFE3, true);
-					setTimeout(rfbWriter.writeKeyEvent, waitTime += 5, true, cc, true);
-					setTimeout(rfbWriter.writeKeyEvent, waitTime += 15, false, 0xFFE3, true);
-					setTimeout(rfbWriter.writeKeyEvent, waitTime += 5, false, cc, true);
-					if(useShift){
-						setTimeout(rfbWriter.writeKeyEvent, waitTime += 45, false, 0xFFE3, true);
-						setTimeout(rfbWriter.writeKeyEvent, waitTime += 5, false, 0xFFE1, true);
-					}
-					waitTime += pastePauseDelay;
-				}
-				if(input.length > 1) {
-					setTimeout(function() : void {captureKeyEvents = true}, waitTime);
+
+				var i:Number = 0;
+				for (i = 0; i<input.length ;i++) {
+					var char:String = input.charAt(i)
+					var charCode:Number = input.charCodeAt(i);
+					var useShift:Boolean = shiftedChars.indexOf(char) >= 0;
+					var useControl:Boolean = false;
+					charsToSend.push({code: charCode, shifted: useShift, controled: useControl});
 				}
 
+				setTimeout(sendCharsFromQueue, 0);
```

Comment:
to let the function return first, before going all crazy on pasting, but tbh it probably doesn't matter